### DISCUSSION TITLE
Differentiate when using IDs to break circular relation

### DISCUSSION THREAD

**Discussion by fl034** [strapi/strapi#4891] [Legacy Topics]
_I didn't put this on the product board yet, because I want to get other opinions first._

When having circular relations Strapi breaks the circle by replacing the related object by only its ID.

Sometimes it's like this:

```
{ "parent": <ParentObject> }
```

And sometimes like that:

```
{ "parent": "id123" }
```

This behaviour makes it difficult to parse for some more type-safe languages like Swift or Kotlin.

For example in Swift you define a model and make it conform to Codable to enable JSON-parsing for free. But if one property can be either an Object of a specific related type OR a String (ID) we need to write a lot of code or have two models for those two cases.

My suggestion would be to always provide two JSON properties when the API is being consumed. If the relation is called `parent` then it could look like this:

```
{
  "parent": { <ParentObject> },
  "parentId": "123",
}
```

The `parent` could be `null` when Strapi needs to break circular relation.

**Comment by derrickmehaffy** [comment]
The difference is typically when you start nesting to a certain level, by default Strapi only goes 1 level deep (though you can custom populate more levels down in a relation tree). While I do understand what you are requesting, I don't think adding the the extra field would really provide much benefit, you would be repeating the same data in the `parent: object` since you would want to have the ID in there also. then you start getting into other relations such as manyToMany where `parentId` is going to be an array.

I'll mark this for discussion although it didn't properly follow one of the required templates, in order to get further discussion on it.

**Comment by fl034** [comment]
> I don't think adding the the extra field would really provide much benefit, you would be repeating the same data in the parent: object since you would want to have the ID in there also.

Sure, but this is what happens today anyway, right? Just in cases where we nest the object directly, we'd have both fields populated `parent` and `parentId`.

> then you start getting into other relations such as manyToMany where parentId is going to be an array.

Here we could have an array of IDs alongside an array of nested objects, simple :).

> by default Strapi only goes 1 level deep (though you can custom populate more levels down in a relation tree)

Is this somewhere documented or could you lead me to the place in code? Would be cool to override this number for cases where 2 levels are just what one needs.

**Comment by derrickmehaffy** [comment]
> Is this somewhere documented or could you lead me to the place in code? Would be cool to override this number for cases where 2 levels are just what one needs.

https://github.com/strapi/strapi/issues/4062

(appears the docs links are broken in that issue latest links are: https://strapi.io/documentation/3.0.0-beta.x/concepts/services.html)

As to your other comments, you are more than welcome to do this on your own project but I don't think it's something that should be done globally for all Strapi users. Having the same data repeated twice in the response is only going to increase the payload size and thus slow down upstream applications. (Especially mobile devices on slow connections). The goal should be to decrease payload and response time.

**Comment by fl034** [comment]
Thanks for the links!

Okay, keeping payload size in mind, I could reduce my feature-request to the following:

**`Id` is being appended to the JSON field name in those cases when the nested data is being replaced by the ID of the object.**

That way, developers in type safe languages always know what type to expect.

I can try to do that in a PR with a setting to enable/disable it.

**Comment by derrickmehaffy** [comment]
can you give an example of what your expected output would look like compared to the current implementation?

**Comment by fl034** [comment]
**Current implementation** with nested object

```
{ "manufacturer": { <Manufacturer Object> } }
```

**Current implementation:** When Strapi replaces nested object (when more than 1 level of nesting, or to break circular relation) with the ID.

```
{ "manufacturer": "5dbb344857e0e70752bee0a0" }
```

**My suggestion:** When Strapi replaces nested object (when more than 1 level of nesting, or to break circular relation), it appends `Id` to the key of the key-value pair.

```
{ "manufacturerId": "5dbb344857e0e70752bee0a0" }
```

**Comment by derrickmehaffy** [comment]
🤔 To be frank, I think this would cause more confusion as doing nested queries would yield a different result. At the client you could easily just do a `typeof` to know if you are looking at an object or a string/integer (based on the database being used)
